sql: warn users that sequences are slower than using UUID

biradarganesh25 opened the pull request:
Even though the documentation tells the user to use
UUIDs for auto-generating unique IDs, many users might
not read the docs. This warning  will make users aware of
better alternatives than using sequences.

Release note (sql change): Warn users that sequences are
slower than using UUID.

Fixes #41258

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68964)
<!-- Reviewable:end -->

biradarganesh25 commented:
Hey @rafiss could you take a look? I've followed your advice on the issue. I have a feeling I am not checking for the "nextva" properly, but couldn't think of any another way. Lemme know if i gotta change it, thanks

knz reviewed (approved):
A few style nits. Leaving correctness review to @rafiss 

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-:-MhDfbGD2C6kwuo8sqS7:bkwrq59)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @awoods187, @biradarganesh25, and @rafiss)

___
*[pkg/sql/create_table.go, line 213 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfH0t38Mm9wisxQ5c:-MhDfH0t38Mm9wisxQ5d:b3dc7jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/create_table.go#L213)):*
> ```Go
> }
> func hasPrimaryKeySerialType(params runParams, colDef *tree.ColumnTableDef) (bool, error) {
> ```

nit: remove empty line
___
*[pkg/sql/create_table.go, line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfIqO7rkg3rMXeJOb:-MhDfIqO7rkg3rMXeJOc:b3dc7jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/create_table.go#L242)):*
> ```Go
> 
> 	for _, def := range n.n.Defs {
> ```

nit: remove empty line
___
*[pkg/sql/create_table.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfK8TAfHuKnb4wtut:-MhDfK8TAfHuKnb4wtuu:b3dc7jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/create_table.go#L246)):*
> ```Go
> 		case *tree.UniqueConstraintTableDef:
> 			if v.PrimaryKey {
> ```

nit: remove empty line
___
*[pkg/sql/create_table.go, line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfMVQ6nAkxbKxVmAO:-MhDfMVQ6nAkxbKxVmAP:b-c0x80j) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/create_table.go#L251)):*
> ```Go
> 				}
> 			}
> 		case *tree.ColumnTableDef:
> ```

nit: you can add an empty line above `case` here
___
*[pkg/sql/create_table.go, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfLHL9gMn1vI01aKC:-MhDfLHL9gMn1vI01aKD:b3dc7jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/create_table.go#L255)):*
> ```Go
> 				colsWithPrimaryKeyConstraint[v.Name] = true
> 			}
> ```

nit: remove empty line
___
*[pkg/sql/logictest/testdata/logic_test/create_table, line 488 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfWW796t1y35i9mdZ:-MhDfWW796t1y35i9md_:b-m9scuf) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/logictest/testdata/logic_test/create_table#L488)):*
> ```
> 
> query T noticetrace
> CREATE TABLE serial_test2 (id SERIAL PRIMARY KEY, temp string )
> ```

nit: stray space at end
___
*[pkg/sql/logictest/testdata/logic_test/create_table, line 493 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfXmS9xQUSzQ_3gxP:-MhDfXmS9xQUSzQ_3gxQ:bn9j1fx) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/logictest/testdata/logic_test/create_table#L493)):*
> ```
> 
> query T noticetrace
> CREATE TABLE serial_test3 (id SERIAL , temp string, PRIMARY KEY (id))
> ```

nit: stray space before comma


<!-- Sent from Reviewable.io -->

biradarganesh25 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68964)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @awoods187, @knz, and @rafiss)

___
*[pkg/sql/create_table.go, line 213 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfH0t38Mm9wisxQ5c:-MhNvFV87av4F9c9QhCa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/create_table.go#L213)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: remove empty line
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfIqO7rkg3rMXeJOb:-MhNvGu1BX_aZnV034wy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/create_table.go#L242)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: remove empty line
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 246 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfK8TAfHuKnb4wtut:-MhNvHo2DQVJNpqNKJ0E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/create_table.go#L246)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: remove empty line
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 251 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfMVQ6nAkxbKxVmAO:-MhNvIVf1UQS-aKCh3_s:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/create_table.go#L251)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: you can add an empty line above `case` here
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfLHL9gMn1vI01aKC:-MhNvJ7c8FDTDKsXh5eJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/create_table.go#L255)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: remove empty line
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/create_table, line 488 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfWW796t1y35i9mdZ:-MhNvens3_ePT-3Kcu0N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/logictest/testdata/logic_test/create_table#L488)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: stray space at end
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/create_table, line 493 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhDfXmS9xQUSzQ_3gxP:-MhNvlArCAvfzVJeTlAD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/798df18ace16cd80a9d5dcc3f2009676dbb8f17e/pkg/sql/logictest/testdata/logic_test/create_table#L493)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: stray space before commit
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

biradarganesh25 commented:
Thanks @knz! I've updated the file. Also, I realized I had not synced with upstream before raising PR, which is why a lot of CI test failed (I'm guessing). I've updated that also. Got a small doubt: I'd like to run only the create_table logictest but this command is failing as ```no test files found```: ```make test PKG=./pkg/sql/logictest TESTS=TestLogic TESTFLAGS="-config local -d create_table"```
If you see something obviously wrong in the command please let me know, the entire logictest takes quite a bit of time to run. Thanks!

biradarganesh25 commented:
Ah never mind, you need to give full path. Just figured it out.

knz commented:
Try this, much simpler and faster:

```
make testbaselogic FILES=create_table TESTCONFIG=local
```

rafiss reviewed (changes requested):
looks pretty  good! i have small requests, and also your change needs to be rebased on top of master to address a merge conflict.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-:-MhnnpQbC7jq1k7OZHkz:b-oiw3y2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @awoods187, @biradarganesh25, and @knz)

___
*[pkg/sql/create_table.go, line 215 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhnnP7w0hShtOxerz57:-MhnnP7w0hShtOxerz58:bli4ywh) ([raw file](https://github.com/cockroachdb/cockroach/blob/66e3bae0aa82e77ac1507a63f28b2008dffcca5d/pkg/sql/create_table.go#L215)):*
> ```Go
> 	if colDef.IsSerial {
> 		return true, nil
> 	}
> ```

can you add one more case here? 

```
	if colDef.IsSerial || colDef.GeneratedIdentity.IsGeneratedAsIdentity {
		return true, nil
	}
```

(generated as identity is a newer feature that is essentially an alias for SERIAL)
___
*[pkg/sql/create_table.go, line 224 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-Mhnn6LGALYzWFx8_rxL:-Mhnn6LGALYzWFx8_rxM:b-vmxl5e) ([raw file](https://github.com/cockroachdb/cockroach/blob/66e3bae0aa82e77ac1507a63f28b2008dffcca5d/pkg/sql/create_table.go#L224)):*
> ```Go
> 			name = t.Name
> 		case *tree.UnresolvedName:
> 			name = t.String()
> ```

this case should be

```
		case *tree.UnresolvedName:
			fn, err := t.ResolveFunction(params.SessionData().SearchPath)
			if err != nil {
				return false, err
			}
			name = fn.Name
```
___
*[pkg/sql/create_table.go, line 266 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhkKOcqCmz3GQQQz5CU:-MhkKOcqCmz3GQQQz5CV:bpb1byg) ([raw file](https://github.com/cockroachdb/cockroach/blob/66e3bae0aa82e77ac1507a63f28b2008dffcca5d/pkg/sql/create_table.go#L266)):*
> ```Go
> 
> 				if primaryKeySerial {
> 					params.p.BufferClientNotice(
> ```

sending the notice should be done outside of the loop -- we only want to send one notice per create-stmt


<!-- Sent from Reviewable.io -->

biradarganesh25 committed:
sql: warn users that sequences are slower than using UUID

Even though the documentation tells the user to use
UUIDs for auto-generating unique IDs, many users might
not read the docs. This warning  will make users aware of
better alternatives than using sequences.

Release note (sql change): Warn users that sequences are
slower than using UUID.

Fixes #41258

biradarganesh25 reviewed (commented):
Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-:-MjM_jBnAC_oc-a-ebw-:b-896fix)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @awoods187, @knz, and @rafiss)

___
*[pkg/sql/create_table.go, line 215 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhnnP7w0hShtOxerz57:-MjM_kpg9CcdDRxks7E-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66e3bae0aa82e77ac1507a63f28b2008dffcca5d/pkg/sql/create_table.go#L215)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

can you add one more case here? 

```
	if colDef.IsSerial || colDef.GeneratedIdentity.IsGeneratedAsIdentity {
		return true, nil
	}
```

(generated as identity is a newer feature that is essentially an alias for SERIAL)
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 224 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-Mhnn6LGALYzWFx8_rxL:-MjM_mAb3UoOClf3sPc9:b-j3vife) ([raw file](https://github.com/cockroachdb/cockroach/blob/66e3bae0aa82e77ac1507a63f28b2008dffcca5d/pkg/sql/create_table.go#L224)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this case should be

```
		case *tree.UnresolvedName:
			fn, err := t.ResolveFunction(params.SessionData().SearchPath)
			if err != nil {
				return false, err
			}
			name = fn.Name
```
</blockquote></details>

I knew I had messed up here - wasn't sure where to get the serach path, thanks. Made the change.
___
*[pkg/sql/create_table.go, line 266 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhkKOcqCmz3GQQQz5CU:-MjM_vTwD-4FbLL1UJ9x:b-ihykrv) ([raw file](https://github.com/cockroachdb/cockroach/blob/66e3bae0aa82e77ac1507a63f28b2008dffcca5d/pkg/sql/create_table.go#L266)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

sending the notice should be done outside of the loop -- we only want to send one notice per create-stmt
</blockquote></details>

There's a break after the notice is added (inside the if condition itself).... would that be enough? 


<!-- Sent from Reviewable.io -->

biradarganesh25 reviewed (commented):
Also, sorry for leaving the PR for too long, too many thing going on (masters). 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-:-MjMaIy3BajSviGtPy3x:b-8aocce)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @awoods187, @knz, and @rafiss)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
no problem, thanks for your work on this! lgtm

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-:-Mj_FYi8837ltp-3eXeV:bvia24w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @awoods187, @biradarganesh25, @knz, and @rafiss)

___
*[pkg/sql/create_table.go, line 266 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/68964#-MhkKOcqCmz3GQQQz5CU:-Mj_FGW-1vRZFFgOBbrk:b-sq16rm) ([raw file](https://github.com/cockroachdb/cockroach/blob/66e3bae0aa82e77ac1507a63f28b2008dffcca5d/pkg/sql/create_table.go#L266)):*
<details><summary><i>Previously, biradarganesh25 (Ganesh Biradar) wrote…</i></summary><blockquote>

There's a break after the notice is added (inside the if condition itself).... would that be enough? 
</blockquote></details>

ah didn't notice that; yeah that's fine


<!-- Sent from Reviewable.io -->
